src/js/plugins/plugin.autoplay.js

```diff
@@ -0,0 +1,127 @@
+/*global BookReader */
+
+/**
+ * Plugin which adds an autoplay feature. Useful for kiosk situations.
+ */
+jQuery.extend(BookReader.defaultOptions, {
+  enableAutoPlayPlugin: true
+});
+
+/**
+ * @override BookReader.setup
+ */
+BookReader.prototype.setup = (function(super_) {
+  return function (options) {
+    super_.call(this, options);
+
+    this.auto      = false;
+    this.autoTimer = null;
+    this.flipDelay = 5000;
+  };
+})(BookReader.prototype.setup);
+
```

Comment:
Oooooh! I think I found the transform that will make these look less awful!

This has the benefits of:

- We can export the `class`, so it can more easily be `import`ed in tests
- Get rid of that ugly function scoping and `super_`
- Get rid of awkward `BookReader.prototype` overrides
- Still a pretty mindless transform

```suggestion
export default class BookReaderWithAutoplayPlugin extends BookReader {
    /** @override */
    setup(options) {
        super.setup(options);

        this.auto = false;
        this.autoTimer = null;
        this.flipDelay = 5000;
    }
}

if (window && window.BookReader) {
    window.BookReader = BookReaderWithAutoplayPlugin;
}
```

Thoughts?

Reply:
Interesting approach! So is the idea then for each plugin, we just keep replacing `window.BookReader` with the class that is getting exported, which should be the most up-to-date version of BookReader (in that it has all other plugins loaded in it)?

Do we need that `window.BookReader` check? I think we can set it regardless.. ie:
```
if (window) {
  window.BookReader = BookReaderWithAutoplayPlugin;
}
```

Reply:
Yeah! That's essentially what it's doing already using es5 syntax. I think you're right, the `window.BookReader` check is unnecessary; unless, do es6 modules imported with `type="module"` have access to `window`? That's the only case where it might matter.

Reply:
So I just thought of something.. since `BookReaderWithAutoplayPlugin` is extending `BookReader`, I think setting `window.BookReader` to that is going to get you just the `BookReaderWithAutoplayPlugin` without any of the other plugins. You're basically overwriting it each time.

Reply:
It would be sort of recursive; since the plugins are loaded in order, each one overwrites the `BookReader` the previous one modified (which is also the way it currently works :P )

e.g.
```js
window.BookReader = class BookReader {}
window.BookReader = class BookReaderWithAutoPlay extends BookReader {}
window.BookReader = class BookReaderWithIframe extends BookReader {}
```

Reply:
Is it extending `window.BookReader` though? I would expect it to be extending the one imported.. ie:
```
import BookReader from './bookreader.js';

class BookReaderWithAutoPlay extends BookReader {}
```

Reply:
In a browser environment it would be extending `window.BookReader`; in a node environment it would be extending the `import`. Not sure what happens in browser modules 🤔

Reply:
Ohhh interesting!

So just thinking out loud (and not a blocker to this)... is there any way we could extend this idea if we wanted to test multiple plugins independent of `window.BookReader`?

Reply:
Hahaha, my thoughts exactly! I've just been trying to see if there's a way to convert a list of subclasses into an inheritance chain...

Reply:
But yeah, I don't think it's blocking. `jest` offers a `window` object, so importing will work as it does in the browser.

Reply:
Actually, just tested doing this for iframe and it requires _a lot_ of polyfills :/ The file went from 2KB min -> 23 Kb :/ We'd need to have some sort of breaking change to include a onetime polyfill otherwise this will be duplicated for every plugin. So let's not do this for now :(

Reply:
Oy! Bummer! Good idea, though!